contracts/Listing.sol

```diff
@@ -1,35 +1,54 @@
 pragma solidity ^0.4.2;
 
 contract Listing {
-  // Addresses represent listing owners and map to an array of listings.
-  mapping (address => string) public listings;
 
   // 0rigin owner
   address public origin;
 
-  // Event denoting that a given address has updated its array of listings
-  // Currently, this means creating or deleting listings
-  // In the future, we will have separate events for specific actions
-  event UpdateListings(address from);
-
   // Defines origin admin address - may be removed for public deployment
   function Listing() {
     origin = msg.sender;
   }
 
-  // Create a new listing
-  function create(string ipfsHash) {
-    // Check for IPFS hash length
-    
-    require(bytes(ipfsHash).length == 46);
+  struct listingStruct {
+    address lister;
+    // Assume IPFS defaults for hash: function:0x12=sha2, size:0x20=256 bits
+    // See: https://ethereum.stackexchange.com/a/17112/20332
+    bytes32 ipfsHash;
+    uint price;
+    uint unitsAvailable;
+  }
 
-    // May want to de-dupe later by checking for existing hash
-    listings[msg.sender] = ipfsHash;
-    UpdateListings(msg.sender);
+  // Array of all listings
+  listingStruct[] public listings;
+
+  // Return number of listings
+  function listingsLength() public constant returns (uint) {
+      return listings.length;
+  }
+
+  // Return listing info
+  function getListing(uint index) public constant returns (uint, address, bytes32, uint, uint) {
+    // TODO: Determine if less gas to do one array lookup into var, and return var struct parts
+    return (
+      index,
+      listings[index].lister,
+      listings[index].ipfsHash,
+      listings[index].price,
+      listings[index].unitsAvailable
+    );
   }
-  
-  // Change this to return multiple listings later
-  function listingForAddress(address _account) public constant returns (string) {
-    return listings[_account];
+
```

Comment:
We haven't set a standard on how to organize smart contracts, but we should probably have some sort of hierarchy for public/private variables, functions, and events. Perhaps move all events above functions rather than interspersing them amongst more complicated code. Thoughts on this?

Reply:
I could go either way, no strong opinion. If events end up pretty tightly bound to functions, then could make sense to define them together. If events are getting triggered from multiple places, would make sense to group them at the top.